extensions/s3/src/main/java/io/deephaven/extensions/s3/S3AsyncClientFactory.java

```diff
@@ -0,0 +1,144 @@
+                        .build())
+                .region(Region.of(instructions.regionName()))
+                .credentialsProvider(instructions.awsV2CredentialsProvider());
+        instructions.endpointOverride().ifPresent(builder::endpointOverride);
+        final S3AsyncClient ret = builder.build();
+        if (log.isDebugEnabled()) {
+            log.debug().append("Creating S3AsyncClient for region: ").append(instructions.regionName()).endl();
+        }
+        return ret;
+    }
+
+    private static class HttpClientConfig {
+        private final int maxConcurrentRequests;
+        private final Duration connectionTimeout;
+
+        HttpClientConfig(final int maxConcurrentRequests, final Duration connectionTimeout) {
+            this.maxConcurrentRequests = maxConcurrentRequests;
+            this.connectionTimeout = connectionTimeout;
+        }
+
+        int maxConcurrentRequests() {
+            return maxConcurrentRequests;
+        }
+
+        Duration connectionTimeout() {
+            return connectionTimeout;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = maxConcurrentRequests;
+            result = 31 * result + connectionTimeout.hashCode();
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final HttpClientConfig that = (HttpClientConfig) o;
+            return maxConcurrentRequests == that.maxConcurrentRequests
+                    && connectionTimeout.equals(that.connectionTimeout);
+        }
+    }
+
+    private static SdkAsyncHttpClient getOrBuildHttpClient(@NotNull final S3Instructions instructions) {
+        final HttpClientConfig config = new HttpClientConfig(instructions.maxConcurrentRequests(),
+                instructions.connectionTimeout());
+        return httpClientCache.computeIfAbsent(config, key -> AwsCrtAsyncHttpClient.builder()
+                .maxConcurrency(config.maxConcurrentRequests())
+                .connectionTimeout(config.connectionTimeout())
+                .build());
+    }
+
+    /**
+     * The following executor will be used to complete the futures returned by the async client. This is a shared
+     * executor across all clients with fixed number of threads. This pattern is inspired by the default executor used
+     * by the SDK ({@code SdkAsyncHttpClient#resolveAsyncFutureCompletionExecutor}).
+     *
+     */
+    private static Executor ensureAsyncFutureCompletionExecutor() {
+        if (futureCompletionExecutor == null) {
+            synchronized (S3AsyncClientFactory.class) {
+                if (futureCompletionExecutor == null) {
+                    futureCompletionExecutor = Executors.newFixedThreadPool(NUM_PROCESSORS, new ThreadFactoryBuilder()
```

Comment:
Could make this configurable, the way we do for `OperationInitializationThreadPool`. But the current approach is probably fine.

Reply:
If we do the usual "-1 means available processors" thing, which we already have at least two places, we're due to introduce a shared utility for interpreting thread counts.